# 스파크 간단히 살펴보기

## 스파크의 기본 아키텍처
- 보통 한 대의 컴퓨터는 대규모 정보를 연산할 만한 자원이나 성능을 가지지 못함
  - 연산을 할 수 있다고 해도 완료하는데 너무 많은 시간이 걸림

- 컴퓨터 클러스터는 <strong>여러 컴퓨터의 자원을 모아 하나의 컴퓨터처럼 사용</strong>할 수 있게 만듦

- 이 <strong>클러스터에서 작업을 조율</strong>할 수 있는 프레임워크가 필요 <strong>-> 스파크</strong>가 그런 역할을 하는 프레임워크

- 스파크가 연산에 사용할 클러스터는 <strong>클러스터 매니저</strong>에서 관리함
  - Standalone
  - 하둡 YARN
  - Mesos

- 작업 처리 과정
  1. 사용자는 클러스터 매니저에 스파크 애플리케이션을 제출(submit)함
  2. 이를 제출받은 클러스터 매니저는 애플리케이션 실행에 필요한 자원을 할당
  3. 할당받은 자원으로 작업 처리
  
<img src="https://t1.daumcdn.net/cfile/tistory/991779375DEA85D00B"/>

## 스파크 애플리케이션

<img src="https://t1.daumcdn.net/cfile/tistory/99916B495DEA927C01"/>

- 스파크 애플리케이션은 <strong>드라이버 프로세스</strong>와 다수의 <strong>익스큐터 프로세스</strong>로 구성
  - 드라이버 프로세스는 클러스터 노드 중 하나에서 실행되며, main()함수 실행
    - 스파크 언어 API를 통해 다양한 언어로 실행 가능
  - 익스큐터 프로세스는 스파크 코드를 실행

- 드라이버 프로세스의 역할
  - 스파크 애플리케이션의 심장 같은 존재로, 애플리케이션 수명 주기 동안 관련 정보를 모두 유지함
  - 사용자 프로그램이나 입력에 대한 응답
  - 전반적인 익스큐터 프로세스의 작업과 관련된 분석, 배포, 스케줄링
  
- 익스큐터 드라이버의 역할
  - 드라이버 프로세스가 할당한 작업을 수행
    - 즉, 할당한 코드를 실행하고, 
    - 진행 상황을 다시 드라이버 노드에 보고

## 스파크의 다양한 언어 API
- 스파크의 언어 API를 이용하면 다양한 프로그래밍 언어로 스파크 코드를 실행할 수 있다.
  - 스파크는 스칼라로 개발
  - 스파크 언어 API를 통해 클러스터 머신에서 실행되는 스파크 코드로 변환
  
- 스파크는 사용자를 대신해 <strong>파이썬이나 R로 작성한 코드를 익스큐터의 JVM에서 실행할 수 있는 코드로 변환함</strong>

## SparkSession
- 스파크 애플리케이션은 <strong>SparkSession이라 불리는 드라이버 프로세스로 제어</strong>한다.
- SparkSession 인스턴스는 사용자가 정의한 처리 명령을 클러스터에서 실행한다.
- 하나의 SparkSession은 하나의 스파크 애플리케이션에 대응한다.

In [0]:
#spark 변수로 SparkSession 사용 가능
spark

SparkSession - hive 
 
 
 SparkContext 

 Spark UI 

 
 Version 
 v2.4.3 
 Master 
 local[8] 
 AppName 
 Databricks Shell

In [0]:
myRange = spark.range(1000).toDF('number')

In [0]:
display(myRange)

number 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233 234 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251 252 253 254 255 256 257 258 259 260 261 262 263 264 265 266 267 268 269 270 271 272 273 274 275 276 277 278 279 280 281 282 283 284 285 286 287 288 289 290 291 292 293 294 295 296 297 298 299 300 301 302 303 304 305 306 307 308 309 310 311 312 313 314 315 316 317 318 319 320 321 322 323 324 325 326 327 328 329 330 331 332 333 334 335 336 337 338 339 340 341 342 343 344 345 346 347 348 349 350 351 352 353 354 355 356 357 358 359 360 361 362 363 364 365 366 367 368 369 370 371 372 373 374 375 376 377 378 379 380 381 382 383 384 385 386 387 388 389 390 391 392 393 394 395 396 397 398 399 400 401 402 403 404 405 406 407 408 409 410 411 412 413 414 415 416 417 418 419 420 421 422 423 424 425 426 427 428 429 430 431 432 433 434 435 436 437 438 439 440 441 442 443 444 445 446 447 448 449 450 451 452 453 454 455 456 457 458 459 460 461 462 463 464 465 466 467 468 469 470 471 472 473 474 475 476 477 478 479 480 481 482 483 484 485 486 487 488 489 490 491 492 493 494 495 496 497 498 499 500 501 502 503 504 505 506 507 508 509 510 511 512 513 514 515 516 517 518 519 520 521 522 523 524 525 526 527 528 529 530 531 532 533 534 535 536 537 538 539 540 541 542 543 544 545 546 547 548 549 550 551 552 553 554 555 556 557 558 559 560 561 562 563 564 565 566 567 568 569 570 571 572 573 574 575 576 577 578 579 580 581 582 583 584 585 586 587 588 589 590 591 592 593 594 595 596 597 598 599 600 601 602 603 604 605 606 607 608 609 610 611 612 613 614 615 616 617 618 619 620 621 622 623 624 625 626 627 628 629 630 631 632 633 634 635 636 637 638 639 640 641 642 643 644 645 646 647 648 649 650 651 652 653 654 655 656 657 658 659 660 661 662 663 664 665 666 667 668 669 670 671 672 673 674 675 676 677 678 679 680 681 682 683 684 685 686 687 688 689 690 691 692 693 694 695 696 697 698 699 700 701 702 703 704 705 706 707 708 709 710 711 712 713 714 715 716 717 718 719 720 721 722 723 724 725 726 727 728 729 730 731 732 733 734 735 736 737 738 739 740 741 742 743 744 745 746 747 748 749 750 751 752 753 754 755 756 757 758 759 760 761 762 763 764 765 766 767 768 769 770 771 772 773 774 775 776 777 778 779 780 781 782 783 784 785 786 787 788 789 790 791 792 793 794 795 796 797 798 799 800 801 802 803 804 805 806 807 808 809 810 811 812 813 814 815 816 817 818 819 820 821 822 823 824 825 826 827 828 829 830 831 832 833 834 835 836 837 838 839 840 841 842 843 844 845 846 847 848 849 850 851 852 853 854 855 856 857 858 859 860 861 862 863 864 865 866 867 868 869 870 871 872 873 874 875 876 877 878 879 880 881 882 883 884 885 886 887 888 889 890 891 892 893 894 895 896 897 898 899 900 901 902 903 904 905 906 907 908 909 910 911 912 913 914 915 916 917 918 919 920 921 922 923 924 925 926 927 928 929 930 931 932 933 934 935 936 937 938 939 940 941 942 943 944 945 946 947 948 949 950 951 952 953 954 955 956 957 958 959 960 961 962 963 964 965 966 967 968 969 970 971 972 973 974 975 976 977 978 979 980 981 982 983 984 985 986 987 988 989 990 991 992 993 994 995 996 997 998 999

----------
- 위에서 생성한 DataFrame은 한 개의 column과 1000개의 row로 구성
- 각 row에는 0~999까지의 값이 할당<br>
  - 이 숫자들은 <strong>분산 컬렉션</strong>을 나타냄
  - 숫자 범위의 각 부분이 서로 다른 익스큐터에 할당됨
  - 이게 스파크의 DataFrame

## DataFrame

- 가장 대표적인 구조적 API
- 테이블의 데이터를 <strong>row와 column</strong>으로 단순하게 표현
- <strong>스키마</strong>: column과 column의 타입을 정의한 목록
- 수천 대의 컴퓨터에 분산되어 있음
  - 단일 컴퓨터에 저장하기엔 데이터가 너무 크거나 계산이 너무 오래 걸릴 수 있기에

### 파티션
- 스파크는 <strong>모든 익스큐터가 병렬로 작업을 수행</strong>할 수 있도록 파티션이라 불리는 청크 단위로 <strong>데이터를 분할</strong>함
- 파티션: 클러스터의 물리적 머신에 존재하는 로우의 집합
  - 만약 파티션이 하나라면 스파크에 수천 개의 익스큐터가 있더라도 병렬성은 1
  - 그리고 수백 개의 파티션이 있더라도 익스큐터가 하나라면 병렬성은 1
- DataFrame의 파티션은 실행 중에 데이터가 컴퓨터 클러스터에서 물리적으로 분산되는 방식
- DataFrame을 사용하면 파티션을 수동 혹은 개별적으로 처리할 필요가 없다.
  - 물리적 파티션에 데이터 변환용 함수를 지정하면 스파크가 실제 처리 방법을 결정한다.